stylix/mk-target.nix

```
@@ -0,0 +1,80 @@
+{
+  # name used for creation of options
+  name,
+
+  # name used for description of options
+  humanName,
+
+  # condition for automatically enabling this target
+  autoEnable ? true,
+
+  # options to be added under `options.stylix.targets.${name}`
+  extraOptions ? { },
+
+  # list of functions with args representing Stylix options; function results
+  # only merged if all relevant Stylix options are not null
+  configElements ? [ ],
+
+  # function with args representing Stylix options; function result always
+  # merged
+  generalConfig ? null,
+}:
+# Note: we only return a plain module so that calling code can use module args, while using this as their module body
+let
+  configModule =
+    { config, lib, ... }:
+    let
+      cfg = config.stylix.targets.${name};
+      functionArgNames =
+        fn:
+        lib.pipe fn [
+          lib.functionArgs
+          builtins.attrNames
+        ];
+      getStylixAttrs =
+        fn:
+        lib.genAttrs (functionArgNames fn) (
+          arg:
+          if arg == "cfg" then
+            cfg
+          else if arg == "colors" then
+            config.lib.stylix.colors
+          else
+            config.stylix.${arg}
+              or (throw "stylix: Stylix arg was requested by mk-target but does not exist: ${arg}")
+        );
+      mkConfig = fn: fn (getStylixAttrs fn);
+      mkConditionalConfig =
+        c:
+        if builtins.isFunction c then
+          let
+            attrs = builtins.attrValues (getStylixAttrs c);
+          in
+          lib.mkIf (builtins.all (attr: attr != null) attrs) (mkConfig c)
```

May read better with a named condition?

```suggestion
          let
            attrs = builtins.attrValues (getStylixAttrs c);
            allAttrsNonNull = builtins.all (attr: attr != null) attrs;
          in
          lib.mkIf allAttrsNonNull (mkConfig c)
```

Alt names might be `allAttrsDefined` or `allAttrsExist` or `allAttrsPresent`.

I could nit-pick that being "defined" has its own meaning in the option system. Or that an attr's presence/existence usually means `attrs ? foo` not `attrs.foo != null`. But despite that, those names get the intention across.